src/main/java/org/elasticsearch/script/expression/ExpressionScriptEngineService.java

```diff
@@ -132,8 +142,13 @@ public SearchScript search(Object compiledScript, SearchLookup lookup, @Nullable
                     // TODO: more context (which expression?)
                     throw new ExpressionScriptCompilationException("Field [" + fieldname + "] used in expression must be numeric");
                 }
+
                 IndexFieldData<?> fieldData = lookup.doc().fieldDataService().getForField((NumberFieldMapper)field);
-                bindings.add(variable, new FieldDataValueSource(fieldData));
+                if (methodname == null) {
+                    bindings.add(variable, new FieldDataValueSource(fieldData));
+                } else {
+                    bindings.add(variable, new FieldMethodValueSource(fieldData, methodname));
```

Comment:
The methods available should be based on the field type? e.g. the ones you are adding here should only be available for date fields, and if you try to use it with e.g. a float field, it should give an error like "cannot use getYear() on float field"?